### PR TITLE
CASSNODEJS-12: Deno and Bun support

### PR DESCRIPTION
**Pull request by SiyaoIsHiding**
Example green run: https://github.com/SiyaoIsHiding/nodejs-driver/actions/runs/27187690601

### CODE REVIEW

**Comment by SiyaoIsHiding** [comment]
We want to be informed if the driver does no run on the latest deno/bun, so I think using "latest" instead of hard-coded versions makes sense to me

**Comment by SiyaoIsHiding** [comment]
@toptobes

**Comment by toptobes** [comment]
> We want to be informed if the driver does no run on the latest deno/bun, so I think using "latest" instead of hard-coded versions makes sense to me

I fully agree with the idea but I also don't think it'd be the worst idea to also try to pinpoint and test the minimum compatible bun/deno versions as well just so we have a baseline we can point users to

Also I'm not fully familiar with just how thorough the unit tests are compared to the integration tests but we should maybe at least run the integration tests on deno/bun at least once just to make sure everything still passes?

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -46,6 +46,29 @@ jobs:
...
+      - run: bunx mocha test/unit --recursive -R spec -t 5000
+
+  deno-unit-tests:
+    name: Unit tests on Deno
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: denoland/setup-deno@v2
+        with:
+          deno-version: latest
```
[the remark this answers is not in the document]
Comment (jorgebay) [reply .github/workflows/test.yml:67]: I agree that we can pin it later.

**Review by jorgebay — APPROVED** [review]

**Comment by SiyaoIsHiding** [comment]
I propose we define our platform support policy as follows.
1. Node.js (primary): full support across all active LTS lines (current: 20, 22, 24).
  Failures block releases. (unchanged)
2. Bun & Deno (compatibility-supported): the driver is kept working and CI-tested, but
  only against the latest stable (current: 2.8 for deno and 1.3.14 for bun). No promise for older runtime versions. 

-----

Re: unit tests v.s. integration tests
If we announce platform support on deno & bun, only unit tests is not enough. But integration tests on every PR commit push is too heavy. So, I'd argue we keep the deno & bun unit test in the current PR commit workflow, and I will add another Github Action triggered by every `trunk` commit, that runs deno & bun integration tests with a matrix of supported server versions.

-----

Re: hard-coded versions v.s. `latest`. What kavin says makes sense, and I'll change it to `1.x` and `2.x`.

**Comment by toptobes** [comment]
> Re: hard-coded versions v.s. latest. What kavin says makes sense, and I'll change it to 1.x and 2.x.

oh no I'm actually completely fine with `latest`, I just meant we should also test on a minimum possible version if we can (or at least on a version like 1-3 years old)

**Comment by toptobes** [comment]
> I will add another Github Action triggered by every trunk commit, that runs deno & bun integration tests with a matrix of supported server versions.

Just keep in mind if you do this even if the action fails it'll still be committed to trunk and the PR will still be closed (hence merge queues)

**Comment by SiyaoIsHiding** [comment]
> I just meant we should also test on a minimum possible version if we can (or at least on a version like 1-3 years old)

I know for sure Bun 1.0.6 fails, according to this [PR](https://github.com/apache/cassandra-nodejs-driver/pull/416).
I think we can just claim support of 2.8+ for deno and 1.3.14+ for bun. I don't wanna do a binary search of which versions break the driver.

**Comment by SiyaoIsHiding** [comment]
> Just keep in mind if you do this even if the action fails it'll still be committed to trunk and the PR will still be closed

I think that's fine. That's also how apache/cassandra 's CI works I believe. The possibility of unit tests all pass on deno & bun, but integration tests break on them is very little. And you mentioned configuring the merge queue is painful... I think the pain wouldn't be worth it

**Comment by toptobes** [comment]
> I think we can just claim support of 2.8+ for deno and 1.3.14+ for bun. 

Okay I think this is fine – according to https://endoflife.date/deno, versions before 2.8 are EOL anyways (, and for Bun, it looks like they don't have an EOL policy so it's hard to say what's a good version range to support in the first place

So I'm good w/ just testing on latest then but then in the future do we want to keep 2.8 and 1.3 as baselines for support?

**Comment by toptobes** [comment]
> I think that's fine. That's also how apache/cassandra 's CI works I believe. The possibility of unit tests all pass on deno & bun, but integration tests break on them is very little. And you mentioned configuring the merge queue is painful... I think the pain wouldn't be worth it

Your call here. FWIW I've basically got the art of merge queues down to a very simple science so it would be very quick to set up, so I wouldn't consider complexity a factor.

**Review by toptobes — APPROVED** [review]
Tentatively approved pending your response to above comments. There could defintiely also be some deduplication between the two workflows (via a shared action) but it's not the biggest deal.

**Comment by SiyaoIsHiding** [comment]
> then in the future do we want to keep 2.8 and 1.3 as baselines for support?

No, our policy will be we only support the latest stable release of deno & bun at the time of our driver's release. Let's say, Dec 2026 we release driver 5.0.0, and at that time Deno's latest release is 2.10 and Bun's 1.4.14, then our 5.0.0 will only have official support of Deno 2.10 and Bun 1.4.14, not including any previous versions.

**Comment by toptobes** [comment]
We've decided to not use merge queues for the time being... will seal this with a final LGTM
